_posts/2020-07-21-catalogs.md

```diff
@@ -0,0 +1,188 @@
+
+Frequently, companies start building a data platform with a metastore, catalog, or schema registry of some sort already in place. Those let you clearly separate making the data available from consuming it. That separation has a few benefits:
+
+* **Improved productivity** - The most obvious one. Making data reusable and shifting the focus on building new models/pipelines rather than data cleansing and discovery.
+* **Security** - You can control the access to certain features of the data. For example, you can make the schema of the dataset publicly available, but limit the actual access to the underlying data only to particular teams.
+* **Compliance** - If you have all the metadata in a central entity, it's much easier to ensure compliance with GDPR and similar regulations and legal requirements.
+
+## What is stored in a catalog?
+
+Almost all data sets can be described by certain properties that must be known in order to consume them. Those include:
+
+* **Schema** - It describes the actual contents of the data, what columns it has, what are the constraints (e.g. keys) on which the updates should be performed, which fields can act as time attributes, what are the rules for watermark generation and so on.
+
+* **Location** - Does the data come from Kafka or a file in a filesystem? How do you connect to the external system? Which topic or file name do you use?
+
+* **Format** - Is the data serialized as JSON, CSV, or maybe Avro records?
+
+* **Statistics** - You can also store additional information that can be useful when creating an execution plan of your query. For example, you can choose the best join algorithm, based on the number of rows in joined datasets.
+
+Catalogs don’t have to be limited to the metadata of datasets. You can usually store other objects that can be reused in different scenarios, such as:
+
+* **Functions** - It's very common to have domain specific functions that can be helpful in different use cases. Instead of having to create them in each place separately, you can just create them once and share them with others.
+
+* **Queries** - Those can be useful when you don’t want to persist a data set, but want to provide a recipe for creating it from other sources instead.
+
+## Catalogs support in Flink SQL
+Starting from version 1.9, Flink has a set of Catalog APIs that allows to integrate Flink with various catalog implementations. With the help of those APIs, you can query tables in Flink that were created in your external catalogs (e.g. Hive Metastore). Additionally, depending on the catalog implementation, you can create new objects such as tables or views from Flink, reuse them across different jobs, and possibly even use them in other tools compatible with that catalog. As of Flink 1.11, there are two catalog implementations supported by the community:
+
+  1. A comprehensive Hive catalog
+
+  2. A Postgres catalog (preview, read-only, for now)
```

Comment:
I don't have much prior knowledge about Flink SQL (maybe that makes me the ideal reader for this blog post draft):
Before this post, I though the catalog is just good for storing metadata about tables: I thought you can use the Postgres catalog for storing your Kafka+Avro or Filesystem+Parquet table definitions in Postgres Tables.
After reading this post, it seems that this is possible, but it is also possible to query data from Postgres (because the tables defined in postgres are available in Flink SQL)
Somehow I would expect the "Postgres catalog" to be called "Postgres connector" or "Postgres integration".

Maybe it makes sense to state more explicitly what you can and can not do with these catalogs?


Side question: Why is it Postgres-only, not generically JDBC?

Reply:
Background: I am not the author of the Postgres catalog and I did not directly participate in the design.

The idea of catalogs is that you could do both:
1. store Flink specific metadata
2. query the non-specific external data

Postgres Catalog implements only the latter.  In Flink a connector imo is rather well defined and means either source or sink that you can use for reading/writing data and thus I think it does not fit in here. Integration in my opinion is too broad and not well defined. The purpose of a catalog is to read/write/make use of metadata.

It is Postgres only because, it was implemented that way. I was also surprised when I was writing the blogpost and preparing the demo. I think there is a lot of potential for better unification the current implementation across different DBs.

In the post I tried to give a high level idea why you should think of catalogs when working with SQL and give an overview in a form of an e2e example what you can achieve in Flink. My intention was not to give a comprehensive overview of all available features. Nevertheless I am open for suggestion if you think a differently orientated post would make more sense.

Reply:
I tried to elaborate more explicitly about the two approaches:
* integration
* metadata store

Reply:
Thanks a lot for the explanation!

Maybe it would make sense to add a sentence to the beginning of the blog post that states the purpose of it, for example:
> In this blog post, we want to give a high level overview of catalogs in Flink, and give an overview in the form of an end to end example that you can try out yourself
This would help manage the expectation of the reader. I was somehow expecting a "deep dive" into the new Hive / Postgres catalogs

Reply:
I should add that the blog post is generally good to go as is -- maybe it's just my personal taste, so please don't overthink my feedback.
I don't want to be the guy who comes in last minute and overthrows everything.

Reply:
My personal take is that all feedback is valuable. Moreover I do realize I am not the best writer, but because I want to work on it, feedback can only help. ;)